pkg/sql/importer/import_pgdump.go

```diff
@@ -0,0 +1,581 @@
+					return errors.WithHintf(wrappedErr, "is the target of the db.target form? "+
+						"please use db.schema.target or schema.target format for target in %q", ast.String())
+				}
+			}
+			return wrappedErr
+		}
+	}
+	return nil
+}
+
+// MoveObjectsInTempDatabaseToState is to move the db and the objects in it to
+// a certain state.
+func MoveObjectsInTempDatabaseToState(
+	ctx context.Context, state descpb.DescriptorState, txn *kv.Txn, descsCol *descs.Collection,
+) (*dbdesc.Mutable, []*tabledesc.Mutable, []*schemadesc.Mutable, error) {
+	importedTables := make([]*tabledesc.Mutable, 0)
+	importedSchemas := make([]*schemadesc.Mutable, 0)
+
+	dbDescs, err := descsCol.GetMutableDatabaseByName(
+		ctx,
+		txn,
+		TempImportPGDumpDBName,
+		tree.DatabaseLookupFlags{
+			Required:       true,
+			IncludeOffline: true,
+		})
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	tableDescs, err := descsCol.GetAllTableDescriptorsInDatabase(ctx, txn, dbDescs)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	for _, desc := range tableDescs {
+		mutTableDesc, err := descsCol.GetMutableTableByID(ctx, txn, desc.GetID(), tree.ObjectLookupFlags{
+			CommonLookupFlags: tree.CommonLookupFlags{
+				IncludeOffline: true,
+			},
+		})
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		mutTableDesc.State = state
+		if state == descpb.DescriptorState_OFFLINE {
+			mutTableDesc.OfflineReason = "importing"
+		}
+		importedTables = append(importedTables, mutTableDesc)
+		if err := descsCol.WriteDesc(ctx, false /* kvTrace */, mutTableDesc, txn); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	schemaDescs, err := descsCol.GetAllSchemaDescriptorsInDatabase(ctx, txn, dbDescs)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	for _, desc := range schemaDescs {
+		mutSchemaDesc, err := descsCol.GetMutableSchemaByID(ctx, txn, desc.GetID(), tree.SchemaLookupFlags{
+			IncludeOffline: true,
+		})
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		mutSchemaDesc.State = state
+		if state == descpb.DescriptorState_OFFLINE {
+			mutSchemaDesc.OfflineReason = "importing"
+		}
+		importedSchemas = append(importedSchemas, mutSchemaDesc)
+		if err := descsCol.WriteDesc(ctx, false /* kvTrace */, mutSchemaDesc, txn); err != nil {
+			return nil, nil, nil, err
+		}
+	}
+
+	dbDescs.State = state
+	if err := descsCol.WriteDesc(ctx, false /* kvTrace */, dbDescs, txn); err != nil {
+		return nil, nil, nil, err
+	}
+
+	return dbDescs, importedTables, importedSchemas, err
+}
+
+// ProcessDDLStatements is to create the schema specified with the ddl stmts
+// in the dump file. Specifically, it creates a temporary db, parse the ddl
+// stmts in the dump file and push them to the handler, and use an internal
+// executor to run all these ddls within a single txn.
+// TODO(janexing): Figure out job resumption semantics, via
+// ImportDetail.PgdumpDDLProgress.
+// TODO(janexing): Update job status to reflect stage.
+func ProcessDDLStatements(
```

Comment:
Is it possible to use this for real and not just in unit tests? Otherwise, how do we know if it actually works? If it only works some of the time, is it possible to clean up and fall back to the old import logic (assuming there is one) when this throws an error?